sqlalchemy-stubs/orm/session.pyi

```diff
@@ -32,17 +33,268 @@ from ..sql.base import Options
 
 _T = TypeVar("_T")
 _TSession = TypeVar("_TSession", bound=Session)
+_TSessionTransaction = TypeVar(
+    "_TSessionTransaction", bound=SessionTransaction
+)
+
+_TSharedSessionProtocol = TypeVar(
+    "_TSharedSessionProtocol", bound=_SharedSessionProtocol
+)
+_TSessionTransactionProtocol = TypeVar(
+    "_TSessionTransactionProtocol", bound=_SessionTransactionProtocol
+)
+_TSessionProtocol = TypeVar("_TSessionProtocol", bound=_SessionProtocol)
 
 _BindArguments = Mapping[str, Any]
 
+class _IdentityMap:
+    @overload
+    def __get__(self, instance: None, owner: Any) -> None: ...
+    @overload
+    def __get__(self, instance: object, owner: Any) -> WeakInstanceDict: ...
+
+class _SharedSessionProtocol(Protocol):
+    identity_map: _IdentityMap
+    autoflush: bool
+    @property
+    def dirty(self) -> util.IdentitySet[Any]: ...
+    @property
+    def deleted(self) -> util.IdentitySet[Any]: ...
+    @property
+    def new(self) -> util.IdentitySet[Any]: ...
+    @property
+    def is_active(self) -> bool: ...
+    @property
+    def no_autoflush(
+        self: _TSharedSessionProtocol,
+    ) -> ContextManager[_TSharedSessionProtocol]: ...
+    @util.memoized_property
+    def info(self) -> Mapping[Any, Any]: ...
+    def __contains__(self, instance: Any) -> bool: ...
+    def __iter__(self) -> Iterator[Any]: ...
+    def add(self, instance: Any, _warn: bool = ...) -> None: ...
+    def add_all(self, instances: Any) -> None: ...
+    def expire_all(self) -> None: ...
+    def expire(
+        self, instance: Any, attribute_names: Optional[Iterable[str]] = ...
+    ) -> None: ...
+    def expunge(self, instance: Any) -> None: ...
+    def expunge_all(self) -> None: ...
+    def get_bind(
+        self,
+        mapper: Optional[Any] = ...,
+        clause: Optional[ClauseElement] = ...,
+        bind: Optional[Union[Connection, Engine]] = ...,
+        _sa_skip_events: Optional[Any] = ...,
+        _sa_skip_for_implicit_returning: bool = ...,
+    ) -> Union[Connection, Engine]: ...
+    def is_modified(
+        self, instance: Any, include_collections: bool = ...
+    ) -> bool: ...
+    def in_transaction(self) -> bool: ...
+
+class _SessionProtocol(_SharedSessionProtocol, Protocol):
+    def __enter__(self: _TSessionProtocol) -> _TSessionProtocol: ...
+    def __exit__(self, type_: Any, value: Any, traceback: Any) -> None: ...
+    def begin(
+        self,
+        subtransactions: bool = ...,  # NOTE: Deprecated.
+        nested: bool = ...,
+    ) -> _SessionTransactionProtocol: ...
+    def begin_nested(self) -> _SessionTransactionProtocol: ...
+    def rollback(self) -> None: ...
+    def commit(self) -> None: ...
+    def connection(self) -> Connection: ...
```

Comment:
this is incompatible with the `Session` implementation further down because it takes no arguments.

Reply:
It's completely compatible because `Session.connection()` can be called with no arguments. If more arguments are added to this, then every duck-typed session _must_ implement those arguments.

Reply:
You are right, I failed to notice that all of the connection arguments are optional.  Sorry for the noise.

Reply:
Not a problem. It made me take a second look, which isn't always a bad thing 😄

Reply:
It also helped uncover an edge case in Pyright: https://github.com/microsoft/pyright/issues/1728 :)